pkg/vm/vm.go

```diff
@@ -22,37 +27,352 @@ type VirtualMachineConfig struct {
+	} else {
+		missingVal, err = mem.EmptyMemoryValueAsFelt().Div(dst, knownOpCell.Read())
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	err = unknownOpCell.Write(missingVal)
+	if err != nil {
+		return nil, err
+	}
+
+	return dst, nil
+}
+
+func (vm *VirtualMachine) computeRes(
+	instruction *Instruction, op0Cell *mem.Cell, op1Cell *mem.Cell,
+) (*mem.MemoryValue, error) {
+	if instruction.PcUpdate == Jnz {
+		if instruction.Res == Op1 && instruction.Opcode == Nop && instruction.ApUpdate == AddImm {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("invalid flag combination calculating res")
+	}
+
+	switch instruction.Res {
+	case Op1:
+		return op1Cell.Read(), nil
+	case AddOperands:
+		op0 := op0Cell.Read()
+		op1 := op1Cell.Read()
+		return mem.EmptyMemoryValueAs(op0.IsAddress()).Add(op0, op1)
+	case MulOperands:
+		op0 := op0Cell.Read()
+		op1 := op1Cell.Read()
+		return mem.EmptyMemoryValueAsFelt().Mul(op0, op1)
+	}
+
+	return nil, fmt.Errorf("unknown res flag value: %d", instruction.Res)
+}
+
+func (vm *VirtualMachine) opcodeAssertions(
+	instruction *Instruction,
+	dstCell *mem.Cell,
+	op0Cell *mem.Cell,
+	res *mem.MemoryValue,
+) error {
+	switch instruction.Opcode {
+	case Call:
+		// Store at [ap] the current fp
+		err := dstCell.Write(mem.MemoryValueFromSegmentAndOffset(executionSegment, vm.Context.Fp))
+		if err != nil {
+			return err
+		}
+
+		// Write in [ap + 1] the instruction to execute
+		err = op0Cell.Write(
+			mem.MemoryValueFromSegmentAndOffset(programSegment, vm.Context.Pc+uint64(instruction.Size())),
+		)
+		if err != nil {
+			return err
+		}
+
+	case AssertEq:
+		// assert that the calculated res is stored in dst
+		err := dstCell.Write(res)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (vm *VirtualMachine) updatePc(
+	instruction *Instruction,
+	dstCell *mem.Cell,
+	op1Cell *mem.Cell,
+	res *mem.MemoryValue,
+) (uint64, error) {
+	switch instruction.PcUpdate {
+	case NextInstr:
+		return vm.Context.Pc + uint64(instruction.Size()), nil
+	case Jump:
+		return res.Uint64()
+	case JumpRel:
+		relAddr, err := res.Uint64()
+		if err != nil {
+			return 0, err
+		}
+		return vm.Context.Pc + relAddr, nil
+	case Jnz:
+		dest, err := dstCell.Read().ToFieldElement()
+		if err != nil {
+			return 0, err
+		}
+
+		if dest.IsZero() {
+			return vm.Context.Pc + uint64(instruction.Size()), nil
+		}
+
+		// todo(rodro): math check when relAddr is negative
+		relAddr, err := res.Uint64()
+		if err != nil {
+			return 0, err
+		}
+		return vm.Context.Pc + relAddr, nil
+
+	}
+	return 0, fmt.Errorf("unkwon pc update value: %d", instruction.PcUpdate)
+}
+
+func (vm *VirtualMachine) updateAp(instruction *Instruction, res *mem.MemoryValue) (uint64, error) {
+	if instruction.Opcode == Call {
+		if instruction.ApUpdate == Add2 {
+			return vm.Context.Ap + 2, nil
+		}
+		return 0, fmt.Errorf("cannot update ap, invalid flag combination: Call without Add2")
```

Comment:
Should we probably move ```res_logic == 0 && opcode == 0 && ap_update != 1``` logic entirely into decode instruction method? This has nothing to do with the execution of an instruction